app/src/main/java/com/hedvig/app/feature/profile/data/ObserveProfileUiStateUseCase.kt

```diff
@@ -0,0 +1,24 @@
+package com.hedvig.app.feature.profile.data
+
+import arrow.core.Either
+import com.hedvig.android.owldroid.graphql.ProfileQuery
+import com.hedvig.app.feature.profile.ui.tab.ProfileQueryDataToProfileUiStateMapper
+import com.hedvig.app.feature.profile.ui.tab.ProfileUiState
+import com.hedvig.app.util.apollo.QueryResult
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.map
+import kotlinx.coroutines.flow.mapLatest
+
+class ObserveProfileUiStateUseCase(
+    private val profileRepository: ProfileRepository,
+    private val profileQueryDataToProfileUiStateMapper: ProfileQueryDataToProfileUiStateMapper,
+) {
+    suspend fun invoke(): Flow<Either<QueryResult.Error, ProfileUiState>> {
+        return profileRepository
+            .profile()
+            .map(QueryResult<ProfileQuery.Data>::toEither)
+            .mapLatest { either ->
+                either.map { profileQueryDataToProfileUiStateMapper.map(it) }
+            }
+    }
+}
```

Comment:
So about this one. The only reason I added this to be between the VM and the Repository is so that I wouldn't make the Repository have to do the dto -> domain model mapping, but have it serve the data coming from `profile()` as it was before, and have the mapping being done here.
With that said, I don't know what's the "best" way to do this, or simply what it is that we prefer in such cases, so I am completely open to changing this if you have some opinions.
We could change it so that the repository changes the `profile()` function to return a flow of the mapped type directly, so `ProfileUiState` instead of `ProfileQuery.Data`. I am open to either way.

Reply:
Looks sensible to me.